substrate/frame/revive/src/migrations/v1.rs

```diff
@@ -0,0 +1,150 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Multi-Block Migration v1
+//!
+//! This migrate the old `ContractInfoOf` storage to the new `AccountInfoOf`.
+
+extern crate alloc;
+
+use super::PALLET_MIGRATIONS_ID;
+use crate::{weights::WeightInfo, AccountInfo, AccountInfoOf, Config, H160};
+use frame_support::{
+	migrations::{MigrationId, SteppedMigration, SteppedMigrationError},
+	pallet_prelude::PhantomData,
+	weights::WeightMeter,
+};
+
+#[cfg(feature = "try-runtime")]
+use alloc::collections::btree_map::BTreeMap;
+
+#[cfg(feature = "try-runtime")]
+use alloc::vec::Vec;
+
+/// Module containing the old storage items.
+pub mod old {
+	use super::Config;
+	use crate::{pallet::Pallet, ContractInfo, H160};
+	use frame_support::{storage_alias, Identity};
+
+	#[storage_alias]
+	/// The storage item that is being migrated from.
+	pub type ContractInfoOf<T: Config> = StorageMap<Pallet<T>, Identity, H160, ContractInfo<T>>;
+}
+
+/// Migrates the items of the [`old::ContractInfoOf`] map into [`crate::AccountInfoOf`].
+pub struct Migration<T: Config>(PhantomData<T>);
+
+impl<T: Config> SteppedMigration for Migration<T> {
+	type Cursor = H160;
+	type Identifier = MigrationId<17>;
+
+	fn id() -> Self::Identifier {
+		MigrationId { pallet_id: *PALLET_MIGRATIONS_ID, version_from: 0, version_to: 1 }
+	}
+
+	fn step(
+		cursor: Option<Self::Cursor>,
+		meter: &mut WeightMeter,
+	) -> Result<Option<Self::Cursor>, SteppedMigrationError> {
+		let required = <T as Config>::WeightInfo::v1_migration_step();
+		if meter.remaining().any_lt(required) {
+			return Err(SteppedMigrationError::InsufficientWeight { required });
+		}
+
+		let mut iter = if let Some(last_key) = cursor {
+			old::ContractInfoOf::<T>::iter_from(old::ContractInfoOf::<T>::hashed_key_for(last_key))
+		} else {
+			old::ContractInfoOf::<T>::iter()
+		}
+		.drain();
+
+		while let Some((last_key, value)) = iter.next() {
+			AccountInfoOf::<T>::insert(
+				last_key,
+				AccountInfo { account_type: value.into(), ..Default::default() },
+			);
+
+			if meter.try_consume(required).is_err() {
```

Comment:
Should it not consume weight first? Or do you just want to ensure that it always makes progress?

Reply:
I check at the beginning of step before the while loop that there is enough for 1
so after each iteration I check if there is enough for the next one or return early

Reply:
True, yea its correct.